satpy/readers/yaml_reader.py

```diff
@@ -349,6 +349,10 @@ class FileYAMLReader(AbstractYAMLReader, DataDownloadMixin):
 
     """
 
+    # WeakValueDictionary objects must be created at the class level or else
+    # dask will not be able to serialize them on a distributed environment
+    _coords_cache: WeakValueDictionary = WeakValueDictionary()
```

Comment:
I'm wondering if this could be a problem. Are we guaranteed to have only one instance of a FileYAMLReader at any given time?

```python
In [41]: class A:
    ...:     bla = dict()
    ...: 

In [42]: a = A()

In [43]: b = A()

In [44]: a.bla["lons"] = 12

In [45]: b.bla["lons"]
Out[45]: 12
```

Reply:
No we don't have that guarantee, but also I looked at the keys being used and they consist of the `lons.name` of the dask array. So I'm not too concerned about conflict. It is less than ideal, but I figured this was one way to compromise between serializable and cache-able. Definitely open to other ideas. Like maybe a regular dictionary as an instance cache?

Reply:
Ok, so let's take a step back here: `_make_swath_definition_from_lons_lats` could be made into a function if we pass `self.info["sensors]` to it, or even pass it `sensor_str` as a general prefix for example.
From there we could implement a small caching decorator using a WeakValueDict:
```python
def weak_cache(function):
  cache = WeakValueDictionary()
  def wrapper(*args):
    try:
      return cache[args]
    else:
      res = function(*args)
      cache[args] = res
      return res
  return wrapper
```
so that we could write
```python
@weak_cache
def _make_swath_definition_from_lons_lats(lons, lats, prefix):
    ...
```

What do you think? would it work?

Reply:
1. To clarify, this does not differ functionally from the class cache, right? It is still a single cache for all reader instances.
2. The decorator couldn't be a generic function like that because `args` would have to be hashable and the existing `_make_swath_definition_from_lons_lats` only uses the dask name in the cache key: `(lons.data.name, lats.data.name)`. So to handle that we'd have to probably make this decorator specific to this function. Plus we'd be added the sensors argument so that complicates what should be used in the cache key.

Reply:
SRP of the method? Or SRP of the class? Either way, class cache (this PR) versus instance cache (main) doesn't change that.

Reply:
True. I made my peace with the class cache, I was just trying to make it overall cleaner.

Reply:
> I was just trying to make it overall cleaner.

How dare you. But really, I see what you're saying but maybe we save that kind of thing for a base reader refactor session in the future. I'll merge this in a little bit.